ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,765 @@
+            .is_some_and(|x| !x.lock().unwrap().is_empty())
+    }
+}
+
+/// Steps used by the validation component of the DZKP
+#[cfg_attr(feature = "descriptive-gate", derive(ipa_macros::Step))]
+pub(crate) enum Step {
+    /// For the execution of the malicious protocol.
+    DZKPMaliciousProtocol,
+    /// Step for validating the DZK proof.
+    DZKPValidate,
+}
+
+/// Marker Trait `DZKPBaseField` for fields that can be used as base for DZKP proofs and their verification
+/// This is different from trait `DZKPCompatibleField` which is the base for the MPC protocol
+pub trait DZKPBaseField: Field {
+    type UnverifiedFieldValues;
+    fn convert(
+        x_left: &BitArray<[u8; 32], Lsb0>,
+        x_right: &BitArray<[u8; 32], Lsb0>,
+        y_left: &BitArray<[u8; 32], Lsb0>,
+        y_right: &BitArray<[u8; 32], Lsb0>,
+        prss_left: &BitArray<[u8; 32], Lsb0>,
+        prss_right: &BitArray<[u8; 32], Lsb0>,
+        z_right: &BitArray<[u8; 32], Lsb0>,
+    ) -> Self::UnverifiedFieldValues;
+}
+
+/// Validator Trait for DZKPs
+/// It is different from the validator trait since context outputs a `DZKPUpgradedContext`
+/// that is tied to a `DZKPBatch` rather than an `accumulator`.
+/// Function signature of `validate` is also different, it does not downgrade shares anymore
+#[async_trait]
+pub trait DZKPValidator<B: UpgradableContext> {
+    fn context(&self) -> B::DZKPUpgradedContext;
+
+    /// Allows to validate the current `DZKPBatch` and empties it. The associated context is then
+    /// considered safe until another multiplication is performed and thus new values are added
+    /// to `DZKPBatch`.
+    /// Is generic over `DZKPBaseFields`. Please specify a sufficiently large field for the current `DZKPBatch`.
+    async fn validate<DF: DZKPBaseField>(&self) -> Result<(), Error>;
+
+    /// `is_safe` checks that there are no remaining `UnverifiedValues` within the associated `DZKPBatch`
+    ///
+    /// ## Errors
+    /// Errors when there are `UnverifiedValues` left.
+    fn is_safe(&self) -> Result<(), Error>;
+
+    /// `get_chunk_size` returns the chunk size of the validator
+    fn get_chunk_size(&self) -> Option<usize>;
+
+    /// `seq_join` in this trait is a validated version of `seq_join`. It splits the input stream into `chunks` where
+    /// the `chunk_size` is defined within `validator`. Each `chunk` is a vector that is independently
+    /// verified using `validator.validate()`, which uses DZKPs. Once the validation fails,
+    /// the output stream will return that the stream is done.
+    ///
+    fn seq_join<'st, S, F, O, DF>(
+        &'st self,
+        source: S,
+    ) -> impl Stream<Item = Result<O, Error>> + 'st
+    where
+        S: Stream<Item = F> + Send + 'st,
+        F: Future<Output = O> + Send + 'st,
+        O: Send + Sync + Clone + 'static,
+        DF: DZKPBaseField,
+    {
+        // chunk_size is undefined in the semi-honest setting, set it to 10, ideally it would be 1
```

Comment:
I don't think this optimization gives us anything useful for semi-honest, but makes this code harder to operate/understand. Vote for making chunk size the same in both settings

Reply:
Chunk collects the stream into a vector. Therefore it is a synchronization point but I am fine either way since it is anyway just semi honest security for which we probably dont need to optimize.